pallets/loans-ref/src/valuation.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_primitives::{Moment, SECONDS_PER_YEAR};
+use cfg_traits::ops::{EnsureDiv, EnsureFixedPointNumber, EnsureInto, EnsureMul, EnsureSub};
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+	traits::tokens::{self},
+	RuntimeDebug,
+};
+use scale_info::TypeInfo;
+use sp_arithmetic::traits::checked_pow;
+use sp_runtime::{traits::One, ArithmeticError, FixedPointNumber, FixedPointOperand};
+
+/// Discounted cash flow values
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+#[cfg_attr(test, derive(Default))]
+pub struct DiscountedCashFlow<Rate> {
+	/// The probability of a borrower defaulting a loan repayments.
+	pub probability_of_default: Rate,
+
+	/// The share of an asset that is lost if a borrower defaults.
+	pub loss_given_default: Rate,
+
+	/// Rate of return used to discount future cash flows back to their present value.
+	pub discount_rate: Rate,
+}
+
+impl<Rate: FixedPointNumber> DiscountedCashFlow<Rate> {
+	pub fn new(
+		probability_of_default: Rate,
+		loss_given_default: Rate,
+		discount_rate: Rate,
+	) -> Self {
+		Self {
+			probability_of_default,
+			loss_given_default,
+			discount_rate,
+		}
+	}
+
+	pub fn compute_present_value<Balance: tokens::Balance + FixedPointOperand>(
+		&self,
+		debt: Balance,
+		when: Moment,
+		interest_rate_per_sec: Rate,
+		maturity_date: Moment,
+		origination_date: Moment,
+	) -> Result<Balance, ArithmeticError> {
+		// If the loan is overdue, there are no future cash flows to discount,
+		// hence we use the outstanding debt as the value.
+		if when > maturity_date {
+			return Ok(debt);
+		}
+
+		// Calculate the expected loss over the term of the loan
+		let tel = Rate::saturating_from_rational(
+			maturity_date.ensure_sub(origination_date)?,
+			SECONDS_PER_YEAR,
+		)
+		.ensure_mul(self.probability_of_default)?
+		.ensure_mul(self.loss_given_default)?
+		.min(One::one());
```

Comment:
If `maturity_date` is in the long future, the result could be higher than 1, so I think we want to limit this. You can not check this in config time because the `origination_date` is not set yet.